Create monome-dinome.py

KanavDesh opened the pull request.

KanavDesh committed:
Create monome-dinome.py

AuburnMischief reviewed (changes requested):
Can you please add a CLI function and using if __name__ == "__main__" before I do a full review?

Thanks

KanavDesh committed:
Update monome-dinome.py

lmk if i need to change anything else

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+from typing import List, Dict
+
+def build_keyword_letters(keyword: str, alphabet: List[str]) -> List[str]:
+    seen = set()
```
Specify type using:

seen: Set = set()

However you will have to use **from typing import Set** for this 😅 but you don't have to import typing for lists and dictionaries

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+    while True:
+        try:
+            total_monome_count = int(
+                input("Total number of monomes (including keyword letters): ").strip()
```
take the inputs outside the function (like in a CLI) rather than inside. this will be useful for when we implement autosolvers

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+
+
+def get_monome_digits(monome_letters: List[str]) -> List[str]:
+    monome_numbers = []
```
element types for the list specified pls like:

monome_numbers: list[int] = []

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+            N = input(f"Digit for {L}: ").strip()
+            if N.isdigit() and len(N) == 1 and N not in monome_numbers:
+                monome_numbers.append(N)
+                break
```
using break is very inefficient and will slow down autosolvers in the future. even if it's functional now for manual decryption, it may be very slow to autosolve in the future. so please rewrite this logic to not use break

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+
+def build_keyword_letters(keyword: str, alphabet: List[str]) -> List[str]:
+    seen = set()
+    keyword_letters = []
```
2.5. please specify the type like this:
keyword_letters: list[str] = []

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+            print("Enter a valid number.")
+
+
+def get_monome_letters(keyword_letters: List[str], total_monome_count: int, alphabet: List[str]) -> List[str]:
```
see point 1.5

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+
+def get_omitted_letters(alphabet: List[str]) -> List[str]:
+    print("Enter omitted letters (leave blank for none)")
+    omit_input = input("Omitted letters: ").strip().upper()
```
see point 4

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+            if 1 <= use_row_count <= len(remaining_digits):
+                break
+            print("Enter a valid number.")
+        except ValueError:
```
see point 5

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+            d = input(f"Row {i+1} first-digit: ").strip()
+            if d in remaining_digits and d not in row_headers:
+                row_headers.append(d)
+                break
```
see point 5

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+
+    return dinome_list[:needed]
+
+def build_decode_map(monome_letters: List[str], monome_numbers: List[str], remaining_letters: List[str], dinome_list: List[str]) -> Dict[str, str]:
```
see point 1.5 + line length

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+                if code not in dinome_list and d2 not in monome_numbers:
+                    dinome_list.append(code)
+                if len(dinome_list) >= needed:
+                    break
```
seep point 5

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+    return keyword_letters
+
+
+def get_total_monome_count(minimum: int) -> int:
```
3. can you move all of these CLI functions into the main CLI function as subfunctions instead or just raw code?

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+    return monome_letters
+
+
+def get_monome_digits(monome_letters: List[str]) -> List[str]:
```
see point 1.5 and 3

AuburnMischief commented on ciphers/monome-dinome.py:
```diff
@@ -0,0 +1,185 @@
+    decode_map = build_decode_map(monome_letters, monome_numbers, remaining_letters, dinome_list)
+
+    # PRINT DECODED PLAINTEXT
+    print("Decoded plaintext:")
```
also when making the decode() / solve() function don't print it. print it in the CLI but for the function just return the decoded plaintext

AuburnMischief reviewed (changes requested):
I left a lot of comments and it may seem overwhelming but a lot of them are just 30s fixes. you did an actually good job though.

also can you please include some testing by using pytest

if you need any help just lmk!!! 😁😁😁😁

KanavDesh committed:
Update monome-dinome.py

Idk what you meant by putting them into one decode function so i left each thing as a subfunction anyway lmk if i need to change anything else hopefully all the logic replacing the breaks is fine
